### ISSUE TITLE
[FEATURE] Host local storage support - Experimental 

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#5724]
**Is your feature request related to a problem? Please describe.**

Applications require large volume or high-performance local disks. Live-migration is not a concern.

**Describe the solution you'd like**

As a HCI solution, we can have a build-in local storage support like leveraging LVM or files.
We have experimented lvm-csi-driver in a POC environment.

**Describe alternatives you've considered**

**Additional context**

**Comment by SlavikCA** [comment]
I'm curious: how that feature will be different from simply using `hostPath`?

Is correct to say, that `hostPath` is file-level access and this feature will provide block-level access?

Or is it going to be file-level access, but the disc (volume group?) will be managed by the LVM CSI driver, so it will provide features such mirroring, striping, etc...?

**Comment by Vicente-Cheng** [comment]
> I'm curious: how that feature will be different from simply using `hostPath`?
> 
> Is correct to say, that `hostPath` is file-level access and this feature will provide block-level access?
> 
> Or is it going to be file-level access, but the disc (volume group?) will be managed by the LVM CSI driver, so it will provide features such mirroring, striping, etc...?

Hi @SlavikCA,

This feature will provide the block device.
With LVM, we can easily use various LVM types like mirror and stripe. Also, we would benefit from various LVM features like expansion/snapshot.

**Comment by abonillabeeche** [comment]
FYI - I had the LVM CSI driver installed on 1.3.0 and upgraded to 1.3.1 without an issue. Perhaps worth investigating is the ability to have snapshots and backup, as it is not available out of the box.
<img width="1172" alt="Screenshot 2024-06-19 at 9 51 06 AM" src="https://github.com/harvester/harvester/assets/5146963/880bd273-155a-4b45-aa37-ab8de43517a0">

**Comment by abonillabeeche** [comment]
> Or is it going to be file-level access, but the disc (volume group?) will be managed by the LVM CSI driver, so it will provide features such mirroring, striping, etc...?

Correct. The LVM CSI provides stripping which could be used for large volumes.

**Comment by albinsun** [comment]
Hi Guys,
Per discussed in [HEP PR](https://github.com/harvester/harvester/pull/5956#issuecomment-2228830404), Longhorn plans to support [LVM-based Local Volume](https://github.com/longhorn/longhorn/issues/4935) in `v1.8` (Nov. 2024) or `v1.9` (Mar. 2025) and we will use LH solution for local storage that time.  

If so, why we need to make this feature by ourself now as LH will support it?
Or they are not completely overlapped?

Ref. https://github.com/longhorn/longhorn/wiki/Roadmap#longhorn-v19-march-2025

**Comment by innobead** [comment]
> Hi Guys, Per discussed in [HEP PR](https://github.com/harvester/harvester/pull/5956#issuecomment-2228830404), Longhorn plans to support [LVM-based Local Volume](https://github.com/longhorn/longhorn/issues/4935) in `v1.8` (Nov. 2024) or `v1.9` (Mar. 2025) and we will use LH solution for local storage that time.
> 
> If so, why we need to make this feature by ourself now as LH will support it? Or they are not completely overlapped?
> 
> Ref. [Wiki: Roadmap (longhorn v19 march 2025) (longhorn/longhorn)](https://github.com/longhorn/longhorn/wiki/Roadmap#longhorn-v19-march-2025)

@albinsun it's all about the priority and good question. 

For Longhorn, local volume via LVM is a relatively low priority compared to ongoing v2 development. However, for Harvester, local volume support is an emerging requirement for data-intensive workloads in some cases. This is why we decided to develop an addon first to support an LVM storage driver with limited functions, allowing users to experience this and maximize the capacity of local disks before we implement complete functions such as snapshot, backup, trim, expansion, and encryption, which Longhorn supports. (I know @Vicente-Cheng is also experimenting with snapshot and expansion in the current development, which is good and okay.)

Whether we migrate to the version supported by Longhorn in the future will depend on how it will be used and adopted in Harvester first, and what features are exclusively supported by Longhorn.

cc @Vicente-Cheng @derekbit

**Comment by TachunLin** [comment]
Verifying on `v1.4.0-rc3` 

### Result
<details><summary>$\color{green}{\textsf{PASS}}$ Single node: Use <strong>lvm-csi-driver</strong> storage class volume on VM $~~$</summary>

 * After created the `harvester-csi-driver-lvm`, can enable the `harvester-csi-driver-lvm` on addons page

  ![image](https://github.com/user-attachments/assets/f3a92b00-28ca-45e3-90ba-d2022f7ab32a)

* Can create Storage for LVM volume group on Host 
  ![image](https://github.com/user-attachments/assets/106608a5-a666-4bfd-a41c-5b2e98eb9358)

* Can create Storage Class for LVM provisioner 
  ![image](https://github.com/user-attachments/assets/2970c49b-c5ed-4fcb-be66-8ab6599c0632)

* Check the storage class yaml details contains the `allowedTopologies` on specific node

  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  allowVolumeExpansion: false
  allowedTopologies:
  - matchLabelExpressions:
    - key: topology.lvm.csi/node
      values:
      - hp-164-seeder
  apiVersion: storage.k8s.io/v1
  kind: StorageClass
  metadata:
    creationTimestamp: "2024-10-22T13:38:55Z"
    name: lvm-sc
    resourceVersion: "3969422"
    uid: 12cd92ba-33bb-4e3e-bbb3-44115140a9d7
  parameters:
    type: striped
    vgName: lvm-vg
  provisioner: lvm.driver.harvesterhci.io
  reclaimPolicy: Delete
  volumeBindingMode: WaitForFirstConsumer
  ```


* Can select lvm storage volume and use it as data volume for the VM.

  ![image](https://github.com/user-attachments/assets/08a9cec6-dec8-498d-8859-35ddc7e858c5)

* Check the VM is running and the lvm-sc volume is attached on it

  ![image](https://github.com/user-attachments/assets/27cb8619-b4d1-4058-a0d7-b6f3947392c3)
  
  ![image](https://github.com/user-attachments/assets/d73e7d66-a998-4e09-b6d1-1447f64e6670)
  
  ![image](https://github.com/user-attachments/assets/cb8127ff-6f8a-4f32-8fcb-df534747f661)

 </details>

<details><summary>$\color{red}{\textsf{Failed}}$ Multiple nodes: Use <strong>lvm-csi-driver</strong> storage class volume on VM $~~$</summary>


 * Given a disk on node1 have created by lvm provisioner 
 
  ![image](https://github.com/user-attachments/assets/e85250b1-d520-4622-9509-36c1cb481840)

 * When we try to create the storage class with the `lvm` provisioner type, but can't find the node1 on the available list

  ![image](https://github.com/user-attachments/assets/dcb6be56-0a7c-45b4-bb06-38efcdd7c4ed)

https://github.com/user-attachments/assets/8bd905ae-859d-4898-995e-b166e94e6fc7


* After we create a storage class on node2   

  ![image](https://github.com/user-attachments/assets/62e25467-dba4-4223-8dcc-e324d6a91f7b)


* Then use this storage class to create a data volume in vm 

* The VM would failed to start with `Unschedulable` 

  ```
  0/3 nodes are available: pod has unbound immediate PersistentVolumeClaims. preemption: 0/3 nodes are available: 3 Preemption is not helpful for scheduling.
  ```

  ![image](https://github.com/user-attachments/assets/e0394f7e-9772-4428-a323-cf4b15e2ec92)



  https://github.com/user-attachments/assets/bc510b6a-c918-4fc5-a390-894fb11871bb



 </details>

### Test Information
* Test Environment: Single node and three nodes baremetal machines 
* Harvester version: v1.4.0-rc3

### Verify Steps
<details><summary>Single node: Use <strong>lvm-csi-driver</strong> storage class volume on VM</summary>

1. Create the `harvester-csi-driver-lvm` from yaml in Harvester
  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: Addon
  metadata:
    name: harvester-csi-driver-lvm
    namespace: harvester-system
    labels:
      addon.harvesterhci.io/experimental: "true"
  spec:
    enabled: false
    repo: https://charts.harvesterhci.io
    version: 0.1.4
    chart: harvester-csi-driver-lvm
    valuesContent: ""
  ```

2. Enabled on Add-ons page
  ![image](https://github.com/user-attachments/assets/f3a92b00-28ca-45e3-90ba-d2022f7ab32a)


3. Create Storage for LVM volume group
![image](https://github.com/user-attachments/assets/3d973851-0db8-4d3d-a057-e8d8fef70fff)

4. First time create LVM volume group
  ![image](https://github.com/user-attachments/assets/51a7401a-4a4a-4f4d-b602-74bbe060be7f)
  
  ![image](https://github.com/user-attachments/assets/c12fe8b6-9850-4422-b31a-5d51755d1de9)

4. Create StorageClass, select provisioner LVM 

5. Select node -> Select volume group name -> select volume group type and save
  ![image](https://github.com/user-attachments/assets/106608a5-a666-4bfd-a41c-5b2e98eb9358)

6. Created the storage class of `LVM` provisioner
  ![image](https://github.com/user-attachments/assets/2970c49b-c5ed-4fcb-be66-8ab6599c0632)

7. Check the storage class yaml details contains the `allowedTopologies` on specific node
  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  allowVolumeExpansion: false
  allowedTopologies:
  - matchLabelExpressions:
    - key: topology.lvm.csi/node
      values:
      - hp-164-seeder
  apiVersion: storage.k8s.io/v1
  kind: StorageClass
  metadata:
    creationTimestamp: "2024-10-22T13:38:55Z"
    name: lvm-sc
    resourceVersion: "3969422"
    uid: 12cd92ba-33bb-4e3e-bbb3-44115140a9d7
  parameters:
    type: striped
    vgName: lvm-vg
  provisioner: lvm.driver.harvesterhci.io
  reclaimPolicy: Delete
  volumeBindingMode: WaitForFirstConsumer
  ```

8. Select volume and use it as data volume for the VM.

  ![image](https://github.com/user-attachments/assets/08a9cec6-dec8-498d-8859-35ddc7e858c5)

9. Check the VM is running and the lvm-sc volume is attached on it
  ![image](https://github.com/user-attachments/assets/27cb8619-b4d1-4058-a0d7-b6f3947392c3)
  
  ![image](https://github.com/user-attachments/assets/d73e7d66-a998-4e09-b6d1-1447f64e6670)
  
  ![image](https://github.com/user-attachments/assets/cb8127ff-6f8a-4f32-8fcb-df534747f661)

</details>



<details><summary>Multiple nodes: Use <strong>lvm-csi-driver</strong> storage class volume on VM</summary>

1. Create the `harvester-csi-driver-lvm` from yaml in Harvester
  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: Addon
  metadata:
    name: harvester-csi-driver-lvm
    namespace: harvester-system
    labels:
      addon.harvesterhci.io/experimental: "true"
  spec:
    enabled: false
    repo: https://charts.harvesterhci.io
    version: 0.1.4
    chart: harvester-csi-driver-lvm
    valuesContent: ""
  ```

2. Enabled on Add-ons page
  ![image](https://github.com/user-attachments/assets/115669e3-bc6b-4966-a288-6f13db7c10a0)


3. Create Storage for LVM volume group on node0

  ![image](https://github.com/user-attachments/assets/5f0fed46-f8ad-4d34-b6b9-25bcc73a9414)

4. Create StorageClass, select provisioner LVM 


5. Select node -> Select volume group name -> select volume group type and save

  ![image](https://github.com/user-attachments/assets/c1609c1d-3862-4c43-bf1d-e8c5d40e07c6)

6. Created the storage class of `LVM` provisioner

  ![image](https://github.com/user-attachments/assets/e20c84b1-d539-447d-b479-90f850c846ea)


7. Check the storage class yaml details contains the `allowedTopologies` on specific node

  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  allowVolumeExpansion: false
  allowedTopologies:
  - matchLabelExpressions:
    - key: topology.lvm.csi/node
      values:
      - node2
  apiVersion: storage.k8s.io/v1
  kind: StorageClass
  metadata:
    creationTimestamp: "2024-10-23T08:09:31Z"
    name: lvm-sc2
    resourceVersion: "257300"
    uid: 6d3a3f00-004a-4f1b-be7c-27aaf7c2591a
  parameters:
    type: dm-thin
    vgName: lvm-vg2
  provisioner: lvm.driver.harvesterhci.io
  reclaimPolicy: Delete
  volumeBindingMode: WaitForFirstConsumer
  ```

8. Create vm, select volume and use it as data volume for the VM.



9. Check the VM is running and the lvm-sc volume is attached on it


</details>

**Comment by Vicente-Cheng** [comment]
root cause:
```
WARNING: ext4 signature detected on /dev/sdb at offset 1080. Wipe it?
```
The `vgcreate` failure is due to the target device not being clean.

It should be fixed with that PR: https://github.com/harvester/node-disk-manager/pull/159
You can use v1.4.0-rc4 to test again or wipefs the corresponding device manually.

**Comment by TachunLin** [comment]
Thanks for checking the root cause. 
After using `wipefs` to cleanup the target disk device signature on the node machine. 
```
$ sudo wipefs --all --force /dev/sdc

node2:~ # sudo wipefs --all --force /dev/sdc
/dev/sdc: 2 bytes were erased at offset 0x00000438 (ext4): 53 ef
```

And retest in the following steps:

1. Delete the previously created storage class
2. Remove the /dev/sdc on node2
3. Add the /dev/sdc back on node2 given the lvm-vg2 group already created before
4. Create a new storage class on the disk
5. Create the new vm, add the second volume with the lvm storage class
6. Check the vm is running and can attach the volume well 

The vm can be started in running and attach the lvm volume well
  ![image](https://github.com/user-attachments/assets/56b9f291-ba87-4487-b18c-e6ebbd8d53f5)
  ![image](https://github.com/user-attachments/assets/2be9cbe7-ccc8-413f-8cfe-aba3350653c2)

**Comment by TachunLin** [comment]
Update the retesting state of resting on three nodes `v1.4.0-rc4`

* On a newly provisioned Harvester on kvm machines with extra disks.

* After cleanup the signature with wipefs.
  ```
  rancher@harvester-node-0:~> sudo wipefs /dev/sda
  rancher@harvester-node-0:~> lsblk
  ...
  rancher@harvester-node-0:~> sudo wipefs --all --force /dev/sda
  rancher@harvester-node-0:~>
  ```

* Then adding the lvm provisioner disk on host page. It prompt the error message
  ![image](https://github.com/user-attachments/assets/f4a56284-01ff-45a2-bd97-a989bffb8888)

  ```
  failed to activate volume group lvm-vg1, err: execute command 'vgchange' with args '[--activate y lvm-vg1]' failed: failed to execute: nsenter [--mount=/host/proc/4716/ns/mnt --net=/host/proc/4716/ns/net --ipc=/host/proc/4716/ns/ipc vgchange --activate y lvm-vg1], output , stderr File descriptor 7 (socket:[45102]) leaked on vgchange invocation. Parent PID 1: /usr/lib/systemd/systemd
  Volume group "lvm-vg1" not found
  Cannot process volume group lvm-vg1
  : exit status 5
  ```

* As per discussed with Vicente, it a side-effect we introduce in rc4 with https://github.com/harvester/node-disk-manager/pull/159. 
* We also have the fix PR https://github.com/harvester/node-disk-manager/pull/161.

* The next step is we will verify the add disk again on the `v1.4.0-rc5`.

**Comment by TachunLin** [comment]
Retest **fixed** on `v1.4.0-rc5`. Close this issue.  

### Test Result on single node
<details><summary>$\color{green}{\textsf{PASS}}$ <strong>NVME</strong> disk can be added as lvm-csi-driver disk to Host and be created as volume $~~$</summary>

  1. After enabled `harvester-csi-controller-lvm` addons
  
  2. Can correctly add the `NVME` disk as `LVM` provisioner to the host

  ![image](https://github.com/user-attachments/assets/433abb8c-c378-4014-a5c5-ea35d8887bed)

  3. Can correctly create the storage class with the selected node and volume group
  
  ![image](https://github.com/user-attachments/assets/ccacc394-6229-4ddd-b801-9aac2e6e4080)

  4. VM created with LVM provisioner storage class running well 

  ![image](https://github.com/user-attachments/assets/45d837c5-a847-4a69-b5bf-42eeb868ab12)

  
  5.  Volume create with the LVM provisioner 

  ![image](https://github.com/user-attachments/assets/ba69566a-4d9d-4c96-bc39-b0658b3ae69a)

  6. Can take a snapshot of the create vm
  
  ![image](https://github.com/user-attachments/assets/13cfc2c1-cb8b-4544-9619-8b3c243d70da)

  7. Disk state 

  ```
  v140rc5-test-lvm:~ # lsblk -o NAME,TYPE,TRAN
  NAME                                                                   TYPE TRAN
  loop0                                                                  loop 
  loop1                                                                  loop 
  loop2                                                                  loop 
  ...
  nvme1n1                                                                disk nvme
  ├─lvm--vg1-pvc--812c99b3--a774--4877--8bf5--4bda9fe1edca-real          lvm  
  │ ├─lvm--vg1-pvc--812c99b3--a774--4877--8bf5--4bda9fe1edca             lvm  
  │ └─lvm--vg1-lvm--snapshot--d8215eca--79dd--4650--a4be--7426d643e7d5   lvm  
  └─lvm--vg1-lvm--snapshot--d8215eca--79dd--4650--a4be--7426d643e7d5-cow lvm  
    └─lvm--vg1-lvm--snapshot--d8215eca--79dd--4650--a4be--7426d643e7d5   lvm  
  
  ```

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ <strong>SATA</strong> disk can be added as lvm-csi-driver disk to Host and be created as volume $~~$</summary>

  1. After enabled `harvester-csi-controller-lvm` addons
  
  2. Can correctly add the `SATA` disk as `LVM` provisioner to the host

  ```
  v140rc5-test-lvm:~ # lsblk -o NAME,TYPE,TRAN
  NAME                                                                   TYPE TRAN
  ...
  sdc                                                                    disk   sata
  ```

  ![image](https://github.com/user-attachments/assets/80e2647e-add5-4818-93f8-9210d4c0f52f)


  3. Can correctly create the storage class with the selected node and volume group
  
  ![image](https://github.com/user-attachments/assets/33e0a049-dea2-4fc1-a7b5-a850d4e4eec5)

  4. VM created with LVM provisioner storage class running well 

  ![image](https://github.com/user-attachments/assets/cf88fd3e-af05-47b6-8d56-f8ed3dc435fc)


  5.  Volume create with the LVM provisioner 

  ![image](https://github.com/user-attachments/assets/91619a6e-7165-41e7-b2e6-7c7d00d9c2e6)


  6. Can take a snapshot of the create vm

  ![image](https://github.com/user-attachments/assets/c2e81fe2-a060-4675-9cf5-e32238afa31b)

  
  7. Disk state 

  ```
  v140rc5-test-lvm:~ # lsblk -o NAME,TYPE,TRAN
  NAME                                                                   TYPE TRAN
  ...
  sdc                                                                    disk sata
  ├─lvm--vg2-pvc--3ea4506c--260c--48e1--8528--af4235b765a1-real          lvm  
  │ ├─lvm--vg2-pvc--3ea4506c--260c--48e1--8528--af4235b765a1             lvm  
  │ └─lvm--vg2-lvm--snapshot--2a8acc9d--c8fb--4d55--83bc--bc54552fd97b   lvm  
  └─lvm--vg2-lvm--snapshot--2a8acc9d--c8fb--4d55--83bc--bc54552fd97b-cow lvm  
    └─lvm--vg2-lvm--snapshot--2a8acc9d--c8fb--4d55--83bc--bc54552fd97b   lvm
  ```

 </details>


### Test Result on multiple nodes 

<details><summary>$\color{green}{\textsf{PASS}}$ LVM volume can be created on specific node according to the LVM storage class $~~$</summary>

1. Can display all nodes contains LVM volume group while creating LVM storage class 

  ![image](https://github.com/user-attachments/assets/4cbebf18-2223-43ca-bc06-cf2b4d8801ce)

2. Only the volume group on `n2-140rc5`  would be displayed 
  ![image](https://github.com/user-attachments/assets/bd0de153-6cda-4580-ad9c-4255ff505b81)


3. VM can be created and running ion the node specified in the lvm storage class 

  ![image](https://github.com/user-attachments/assets/6266d1c4-6319-428a-9008-4af744cdc9ff)

  ![image](https://github.com/user-attachments/assets/d1108393-56ef-4ae1-94af-05063269df79)


 </details>



### Test Information
* Test Environment: 
  - Single node harvester on Equinix bare-metal machine (`s3.xlarge.x86`) 
  - Two nodes kvm machines
* Harvester version: `v1.4.0-rc5`

### Verify Steps

<details><summary> Test lvm-csi-driver on single node Harvester </summary>

Refer to [document](https://github.com/Vicente-Cheng/docs/blob/8cde0173d5ccc8bbefe44b4c86a8ff94e55ea99f/docs/advanced/addons/lvm-local-storage.md)

1. Create the `harvester-csi-driver-lvm` from yaml in Harvester
  
  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: Addon
  metadata:
    name: harvester-csi-driver-lvm
    namespace: harvester-system
    labels:
      addon.harvesterhci.io/experimental: "true"
  spec:
    enabled: false
    repo: https://charts.harvesterhci.io
    version: 0.1.4
    chart: harvester-csi-driver-lvm
    valuesContent: ""
  ```

2. Enabled `harvester-csi-driver-lvm` on Add-ons page

  ![image](https://github.com/user-attachments/assets/8c81f6c3-223a-4906-84bb-fcb14b7940ff)


3. Edit Host page to add a disk in Storage 

4. Select NVEM/SATA/SCSI type of disk

6. Select the `LVM` provisioner, and create a volume group `lvm-vg1` 

![image](https://github.com/user-attachments/assets/60563aa7-5d70-447d-a81b-a020b54c59cf)

7. Check the disk can be correct added to the host

![image](https://github.com/user-attachments/assets/433abb8c-c378-4014-a5c5-ea35d8887bed)


8. Create StorageClass, select provisioner LVM 


9. Select node -> Select volume group name -> select volume group type and save

  ![image](https://github.com/user-attachments/assets/c1609c1d-3862-4c43-bf1d-e8c5d40e07c6)

10. Create a new storage class `lvm-sc-nvme`, select `LVM` provisioner

11. Select the `Node`, `Volume Group Name` and `Volume Group Type`

  ![image](https://github.com/user-attachments/assets/fa6997d8-6e12-4619-aa2a-da5168ded858)

12. Check can correctly create the storage class 

  ![image](https://github.com/user-attachments/assets/ccacc394-6229-4ddd-b801-9aac2e6e4080)

13. Create a vm and use the storage class `lvm-sc-nvme` 

  ![image](https://github.com/user-attachments/assets/26d0cf37-ee36-407b-a26c-3c94509dc382)


14. Check the vm can running well 

  ![image](https://github.com/user-attachments/assets/45d837c5-a847-4a69-b5bf-42eeb868ab12)


15. Check the volume is created with the `LVM` provisioner 

  ![image](https://github.com/user-attachments/assets/ba69566a-4d9d-4c96-bc39-b0658b3ae69a)




16. In the csi-driver-config section, select ⋮ > Edit Setting 

17. Add an entry with the following settings and save

  - Provisioner: Select `lvm.driver.harvesterhci.io`.
  - Volume Snapshot Class Name: Select `lvm-snapshot`.
  - Backup Volume Snapshot Class Name: Select `lvm-snapshot`.

  ![image](https://github.com/user-attachments/assets/8d095560-bd3f-4bf5-a965-f3e237c7c03c)



18.  Take a snapshot of the created vm 

19.  Verify that snapshot is ready to us

  ![image](https://github.com/user-attachments/assets/13cfc2c1-cb8b-4544-9619-8b3c243d70da)



</details>



<details><summary> Test lvm-csi-driver on multiple nodes Harvester </summary>

1. 1. Create the `harvester-csi-driver-lvm` from yaml in Harvester
 
2. Enabled `harvester-csi-driver-lvm` on Add-ons page

3. Open the node1 host page -> Storage 

4. Add a disk and select the LVM provisioner, create a new volume group: `lvm-vg1`

![image](https://github.com/user-attachments/assets/3b0d02b3-1230-4428-b3b4-96acdb5aba56)

![image](https://github.com/user-attachments/assets/789268fe-fa43-4f42-a073-baf10dc6a773)


5. Open the node2 host page -> Storage

6. Add a disk and select the LVM provisioner, create a new volume group: `lvm-vg2`

![image](https://github.com/user-attachments/assets/d3284cdd-7738-4117-96b1-272dc6444d9b)

![image](https://github.com/user-attachments/assets/5b3f5159-3f9f-49be-a9c5-9c9b8a4cf894)


7. Create a new storage class `lvm-sc2` with `LVM` provisioner`

8. Select node `n2-140rc5` 

9. Only the volume group on `n2-140rc5`  would be displayed 

  ![image](https://github.com/user-attachments/assets/bd0de153-6cda-4580-ad9c-4255ff505b81)


10.  Create a vm and add a volume, select storage class with `lvm-sc2` 

  ![image](https://github.com/user-attachments/assets/ca1ca0e1-1f1f-4dcc-bedd-ad0bb7fea378)


11. Check the vm can start in running state on node `n2-140rc5` 

  ![image](https://github.com/user-attachments/assets/6266d1c4-6319-428a-9008-4af744cdc9ff)


12.  

</details>
